src/tactic/norm_swap.lean

```diff
@@ -0,0 +1,78 @@
+/-
+Copyright (c) 2021 Yakov Pechersky All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import data.equiv.basic
+import tactic.norm_num
+
+/-!
+# `norm_swap`
+
+Evaluating `swap x y z` for numerals `x y z : ℕ`.
+-/
+
+open equiv tactic expr
+
+/--
+Match a term of shape `app (⇑f) x` and return expressions for:
+  * the type `tf` of `f`
+  * the instance `has_coe_to_fun tf`
+  * `f`
+  * `x`
+  -/
+meta def expr.get_of_coe_fn (e : expr) (f : name) : tactic (expr × expr × expr × expr) :=
+do
+  if e.is_app_of ``coe_fn
+  then do
+    [α, inst, fexpr, x] ← pure e.get_app_args,
+    let fname : name := fexpr.get_app_fn.const_name,
+    if fname = f then pure (α, inst, fexpr, x)
+      else fail!"retrieved function name {fname} is not the expected {f}"
+  else fail "not of coe_fn form with a single argument"
+
+/--
+Creates the application `n c.α p l`, where `p` is a type class instance found in the cache `c`,
+but works on `c.α : Sort*` instead of how `mk_app` uses `c.α : Type*`.
+-/
+meta def tactic.instance_cache.mk_sorted_app (c : instance_cache) (n : name) (l : list expr) :
+  tactic (instance_cache × expr) :=
+do d ← get_decl n,
+   (c, l) ← instance_cache.append_typeclasses d.type.binding_body c l,
+   return (c, (const n [c.univ.succ]).mk_app (c.α :: l))
+
+open norm_num
+
+namespace norm_swap
+
+/--
+A tactic for normalizing `equiv.swap a b c` where `a b c : ℕ` are numerals.
+
+```
+example : equiv.swap 1 2 1 = 2 := by norm_num
+```
+-/
+@[norm_num] meta def eval : expr → tactic (expr × expr) := λ e, do
+  (swapt, coe_fn_inst, f, c) ← expr.get_of_coe_fn e ``equiv.swap,
+  [α, deceq_inst, a, b] ← pure f.get_app_args, -- the swap should have exactly two arguments applied
+  unify α `(ℕ) <|> (fail "currently, norm_swap supports only ℕ"),
+  dic ← mk_instance_cache α,
+  na ← a.to_nat,
+  nb ← b.to_nat,
+  nc ← c.to_nat,
```

Comment:
I think the only reason you're restricting yourself to `nat` is because you use `to_nat` here instead of `to_rat`. This should work on any type the `prove_ne` works on.

Reply:
How hard would it be to make this work on `fin n` too?

Reply:
That's exactly my goal, to make it work for `fin`. I don't think `to_rat` is good enough for it yet, because of the modular arithmetic. I want to make a `norm_fin` extension first that will normalize things that look like `2.succ`. Once that works, then we'll be able to combine them.

Reply:
Indeed, `to_rat` won't get you to `fin` since `norm_num` doesn't prove `fin` disequalities (yet). It does get you to the `int` example, with a couple other superficial changes.